src/lib.rs

```diff
@@ -583,6 +583,40 @@ pub trait Itertools : Iterator {
         tuple_impl::tuple_windows(self)
     }
 
+    /// Return an iterator over all windows, wrapping back to the first
+    /// elements when the window would otherwise exceed the length of the
+    /// iterator, producing tuples of a specific size (up to 4).
+    ///
+    /// `circular_tuple_windows` clones the iterator elements so that they can be
+    /// part of successive windows, this makes it most suited for iterators
+    /// of references and other values that are cheap to copy.
+    ///
+    /// ```
+    /// use itertools::Itertools;
+    /// let mut v = Vec::new();
+    /// for (a, b) in (1..5).circular_tuple_windows() {
+    ///     v.push((a, b));
+    /// }
+    /// assert_eq!(v, vec![(1, 2), (2, 3), (3, 4), (4, 1)]);
+    ///
+    /// let mut it = (1..5).circular_tuple_windows();
+    /// assert_eq!(Some((1, 2, 3)), it.next());
+    /// assert_eq!(Some((2, 3, 4)), it.next());
+    /// assert_eq!(Some((3, 4, 1)), it.next());
+    /// assert_eq!(Some((4, 1, 2)), it.next());
+    /// assert_eq!(None, it.next());
+    ///
+    /// // this requires a type hint
+    /// let it = (1..5).circular_tuple_windows::<(_, _, _)>();
+    /// itertools::assert_equal(it, vec![(1, 2, 3), (2, 3, 4), (3, 4, 1), (4, 1, 2)]);
+    /// ```
+    fn circular_tuple_windows<T>(self) -> CircularTupleWindows<Self, T>
+        where Self: Sized + Clone + Iterator<Item = T::Item> + ExactSizeIterator,
```

Comment:
Are there any issues with eliminating the `ExactSizeIterator` bound?

Reply:
I added `ExactSizeIterator` so that the call to `len()` on line 198 on in `tuple_impl` would be performant. Using `tuple_windows(iter.cycle()).take(len)` was a neat way of getting the extra elements wrapped onto the end, but if removing the `ExactSizeIterator` constraint is preferable we could potentially go with something like the following, at the expense of a slight reduction in readability:
```
pub struct CircularTupleWindows<I, T: Clone>
    where I: Iterator<Item = T::Item> + Clone,
          T: TupleCollect + Clone,
          <T as TupleCollect>::Item: Debug
{
    iter: TupleWindows<Chain<Chain<IntoIter<<T as TupleCollect>::Item>, I>, IntoIter<<T as TupleCollect>::Item>>, T>,
}

pub fn circular_tuple_windows<I, T>(mut iter: I) -> CircularTupleWindows<I, T>
    where I: Iterator<Item = T::Item> + Clone,
          T: TupleCollect + Clone,
          T::Item: Clone,
          <T as TupleCollect>::Item: Debug
{
    let first_elements = iter.by_ref().take(<T as TupleCollect>::num_items() - 1).collect::<Vec<_>>();
    let iter = tuple_windows(first_elements.clone().into_iter().chain(iter).chain(first_elements.into_iter()));

    CircularTupleWindows {
        iter: iter,
    }
}

impl<I, T> Iterator for CircularTupleWindows<I, T>
    where I: Iterator<Item = T::Item> + Clone,
          T: TupleCollect + Clone,
          T::Item: Clone,
          <T as TupleCollect>::Item: std::fmt::Debug
{
    type Item = T;

    fn next(&mut self) -> Option<T> {
        self.iter.next()
    }
}
```

what do you think?